Nouveau protocol abstrait pour implémenter les stratégies

ncs-pl opened the pull request:
- [x] Strategy Pattern
- [x] Nouveaux types pour plus de sécurité dans les jeux (`Choix`, `Plateau`, &c.)
- [x] Protocol abstrait pour la création des jeux dans la classe `Jeu`, permettant d'unifier et simplifier les contrôleurs et les jeux eux-mêmes.
- [x] Unification du processus d'error handling : les jeux throws des erreurs que les contrôleurs ignorent. Seul le `ControlerTemplate` les récupères pour les traiter et afficher. Dans le cas des `CoupInvalideException`, ce même contrôleur demande au joueur de rejouer.
- [x] Stratégies pour les humains

ncs-pl committed:
Model games with abstract classes

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Create game specific Choix input classes

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Move Nim to the new abstract game model

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Migrate Puissance4 to the new model

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Move abstract game engine to a new package

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Create a Jeu abstract protocol

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Make empty abstract classes interfaces

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Complete migration to the new abstract game engine

Signed-off-by: Nicolas Paul <n@nc0.fr>

nc0fr commented:
Le plus impressionnant pour l'instant est de voir le drastique changement et la réduction de code dans les contrôleurs spécifiques, où la méthode `jouerCoup()` est devenue légère !

nc0fr commented:
Il y a aussi le code des jeux eux mêmes qui a évoluer et a été unifiée, il reste encore quelques optimisations possibles que je verrais plus tard (elles n'ont rien d'important)

ncs-pl committed:
Add Strategy interface for gameplayers

Signed-off-by: Nicolas Paul <n@nc0.fr>

YouNes241 commented:
La réduction de code dans le contrôleur est impressionnante !

ncs-pl committed:
Add user's strategy to play Nim

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Add user's strategy to play Puissance4

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Remove casting TODOs

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Make players final

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Remove performance notices

We didn't talk about StringBuilder in class, therefore it's use
is prohibited.

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Simplify methods in IHM

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Remove unused interface

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Update entry point to follow the new flow

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Always make the AI play second

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Display Nim's Tas on their own line

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Fix Nim's state not updating

The issue was caused by JeuNim having its own etatPartie property,
which was updated internally by the game, instead of updating the one
on the Jeu class.
This caused to create a mismatch between the game's state (for Nim, the
game was over) and the controller (which was seeing Jeu's etatPartie,
still being on EtatPartie.EN_COURS).

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Add delay to print functions

Otherwise, stdout and stderr were being written at the same time,
creating inconsistencies in the displayed output within a Terminal
Emulator.

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Change error message on wrong rotation

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Use ASCII characters to display CellulePuissance4

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Make getting invalid cells return null

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Fix end condition checking for draws

In Connect4, we check for a draw if all the cells within the first
line aren't empty.  However, we had reversed the conditional
statement, leading the game to state for a draw from the first move.

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Optimize and fix rotation algorithm

Signed-off-by: Nicolas Paul <n@nc0.fr>

ncs-pl committed:
Add TODO

Need to merge rn

Signed-off-by: Nicolas Paul <n@nc0.fr>
